pykafka/protocol.py

```diff
@@ -794,17 +814,45 @@ def _unpack_message_set(self, buff, partition_id=-1):
         for message in message_set.messages:
             if message.compression_type == CompressionType.NONE:
                 output.append(message)
+                continue
             elif message.compression_type == CompressionType.GZIP:
                 decompressed = compression.decode_gzip(message.value)
-                output += self._unpack_message_set(decompressed,
-                                                   partition_id=partition_id)
+                messages = self._unpack_message_set(decompressed,
+                                                    partition_id=partition_id)
             elif message.compression_type == CompressionType.SNAPPY:
                 decompressed = compression.decode_snappy(message.value)
-                output += self._unpack_message_set(decompressed,
-                                                   partition_id=partition_id)
+                messages = self._unpack_message_set(decompressed,
+                                                    partition_id=partition_id)
+            if messages[-1].offset < message.offset:
+                # With protocol 1, offsets from compressed messages start at 0
+                assert messages[0].offset == 0
+                delta = message.offset - len(messages) + 1
+                for msg in messages:
+                    msg.offset += delta
```

Comment:
Is it safe to artificially adjust the offset values of returned messages? I assume this would result in messages apparently appearing out-of-sequence as offsets jumped around. Imagine getting messages that came from a compressed message set and seeing that their offsets start from 50 and increase. Then after the messages from that messageset have been consumed, an uncompressed message set arrives and again starts returning offsets starting at 51 and increasing. Is that how these offsets work, or am I fundamentally not understanding what's going on here?